include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2012-2017, Association Scientifique pour la Geologie et ses
+ * Applications (ASGA). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of ASGA nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ASGA BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/basic/common.h>
+#include <ringmesh/basic/frame.h>
+
+#include <ringmesh/mesh/mesh.h>
+
+namespace RINGMesh
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( CartesianGridBuilder );
+}
+
+/*!
+ * @file Generic class for creating cartesian grids
+ * @author Melchior Schuh-Senlis
+ */
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API CartesianGrid
+    {
+        ringmesh_disable_copy_and_move( CartesianGrid );
+        friend class CartesianGridBuilder< DIMENSION >;
+
+    public:
+        CartesianGrid( ivecn< DIMENSION > nb_cells_in_each_direction,
+            ReferenceFrame< DIMENSION > vec_cartesian_axis )
+            : nb_cells_in_each_direction_(
+                  std::move( nb_cells_in_each_direction ) ),
+              cartesian_frame_( std::move( vec_cartesian_axis ) ),
+              nb_total_cells_( 1 ),
```

Comment:
What if you have a stupid user and nb_cells_in_each_direction is filled with 0. Your nb_total_cells_ will be equal to 1 which is wrong. You might want to add an exception that capture this case.

Reply:
We multiply nb_total_cells_ by the number of cells in each direction so in this case it will still be zero, but it would be nonsense to have no cell in one direction and cells in another direction, so i'll put something to ensure the number of cells in each direction is >0.